tiledb/sm/serialization/array_schema.cc

```diff
@@ -304,110 +304,139 @@ Status dimension_to_capnp(
   return Status::Ok();
 }
 
-Status dimension_from_capnp(
-    const capnp::Dimension::Reader& dimension_reader,
-    tdb_unique_ptr<Dimension>* dimension) {
-  Datatype dim_type = Datatype::ANY;
-  RETURN_NOT_OK(datatype_enum(dimension_reader.getType().cStr(), &dim_type));
-  dimension->reset(tdb_new(Dimension, dimension_reader.getName(), dim_type));
+tuple<Status, optional<ByteVecValue>> tile_extent_from_capnp(
+    const capnp::Dimension::TileExtent::Reader tile_extent_reader,
+    Datatype dim_type) {
+  auto coord_size = datatype_size(dim_type);
```

Comment:
`dim_type` might be invalid. This code needs to account for that. A check for this needs to be part of a first-line defense against a class of security concerns with insufficient validation of outside data. If this datum is invalid, it needs to be specially logged as such.

At the current time, `datatype_size` returns zero for invalid data. So check for that, document that zero-size is a flag value for "invalid", and leave the function scope. Leaving now should throw. We don't have a specific exception for invalid data, so throw `std::runtime_error("invalid data - <more>")` and identify both the source of the error and the datum that's not valid.

Now there's a related behavior for this below in the `switch` statement, but that statement occurs too late in the function, because derived data (here `coord_size`) is used before `dim_type` has been validated.

Reply:
The latest revision has neither check for validity of `dim_type` nor a message if failed.